### ISSUE TITLE
Python API

### ISSUE THREAD

**Issue by epicfaace** [mxmlnkn/ratarmount#49]
It would be very useful to have a Python API for ratarmount. Specifically, in my use case, I'd like to mount .tar.gz files directly from S3 / Azure Blob Storage -- in order to do that, I would need to be able to pass in file objects to mount (as opposed to actual files) to a ratarmount Python API.

**Comment by mxmlnkn** [comment]
To some extent this should already be possible although it is not documented yet. You can do something like:

```python3
import ratarmount
file = open("foo.tar")
itar = ratarmount.SQLiteIndexedTar(fileObject = file, tarFileName="foo")
print(itar.listDir("/"))
fileInfo = itar.getFileInfo("bar")
print(fileInfo)
print(itar.read("bar", size = fileInfo.size, offset=0))
```

However, as I have no tests for this use case yet, there was a small bug even with this short usage. So, please try with this with the latest development version, onto which I pushed a fix for that:
```
pip install git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount
```

Is this sufficient for your use case?

However, I admit, the usage is quite clunky and it also misses features only available through the TarMount class, which implements the fusepy interface. Here are some possible improvement ideas:

 - Make SQLiteIndexedTar return python file objects to the requested file
 - For returning the file object, hardlinks should be resolved.
 - Maybe extract the union mounting feature into it's own interface so that it can be used programmatically. Currently it is hardcoded into TarMount

**Comment by epicfaace** [comment]
Thanks @mxmlnkn , that's helpful! Here's all what I'm looking to do. I'd like to extract specific files from .tar.gz files on Azure Blob Storage, and list all contents of the archive, without having to seek through / download the entire file, to save network bandwidth. For example, if a .tar.gz file has a 100 GB file and a 1 MB file, reading the 1 MB file should not require one to download the entire .tar.gz archive. I'd also like to use ratarmount as part of a pip package, and thus would rather avoid the dependency on libfuse if possible.

Do you think the approaches I outline below are correct:

- **Creating an index file for a .tar.gz archive** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar`
- **Using an existing index file to read a .tar.gz file** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar`, but pass the path of the index file to `indexFileName` when initializing `SQLiteIndexedTar`.
- **Reading an individual file from a .tar.gz file** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar` as in your code sample above, then use `getFileInfo` and then `read`.
- **Listing the files in a .tar.gz file, or the files within a specified directory within the .tar.gz file** - Do something similar to what you mentioned in your code sample, using `getFileInfo` and then `listDir` (or is it better to use `getFileInfo(listDir=true)` instead?)

I'd be glad to help document some of these use cases / improve the Python API as well!

**Comment by mxmlnkn** [comment]
> I'd also like to use ratarmount as part of a pip package, and thus would rather avoid the dependency on libfuse if possible.

I could try to split off the library from the CLI package. So something like a ratarmount package which depends on ratarmount-lib. And ratarmount-lib would not depend on fuse.

> Do you think the approaches I outline below are correct:
> 
>   * **Creating an index file for a .tar.gz archive** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar`

SQLiteIndexedTar already resolves compression opaquely. So you don't need to use IndexedGzipFile. However, you can do it like outlined if you want, it would work.

>   * **Using an existing index file to read a .tar.gz file** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar`, but pass the path of the index file to `indexFileName` when initializing `SQLiteIndexedTar`.

Yes that works. For indexedGzipFile, you would also need to pass the data for seeking inside the gzip file. Again, SQLiteIndexedTar can do that work for you. It stores this data as a binary blob in the SQLite database.

>   * **Reading an individual file from a .tar.gz file** - I first create an `IndexedGzipFile` and then use `SQLiteIndexedTar` as in your code sample above, then use `getFileInfo` and then `read`.

Yes

>   * **Listing the files in a .tar.gz file, or the files within a specified directory within the .tar.gz file** - Do something similar to what you mentioned in your code sample, using `getFileInfo` and then `listDir` (or is it better to use `getFileInfo(listDir=true)` instead?)

This is how listDir is implemented:
```python3
    def listDir(self, path: str) -> Optional[Iterable[str]]:
        """
        Usability wrapper for getFileInfo(listDir=True) with FileInfo stripped if you are sure you don't need it.
        """
        result = self.getFileInfo(path, listDir=True)
        if isinstance(result, dict):
            return result.keys()
        return None
```
As you can see, getFileInfo(listDir=true) does not only return file names but also information for each file like file size etc.

**Comment by epicfaace** [comment]
Thanks! That's really helpful. I'll proceed with my implementation and try to raise issues / make PRs to fix issues that I encounter.

**Comment by epicfaace** [comment]
@mxmlnkn I'm having some issues with using `.read()` when passing in a fileobj -- it appears to give me garbled data. Looking further, it appears that IndexedGzipFile doesn't support file objects as it expects them to have a fileno / file descriptor...

**Comment by mxmlnkn** [comment]
`indexed_bzip2` will have the same issue then. It also expects at least a fileno. I can't simply forward a Python object to C/C++ code. That's why it uses fileno. I might be able to extend `indexed_bzip2` to work with data in memory, i.e., a simple `char*` but then you would most likely require the whole file to be in memory. It might be possible to use python objects in the C/C++ code, but then you would have to migrate the nice and short Cython code to pure low-level Python C API. This is currently out of scope for me, sorry.

**Comment by epicfaace** [comment]
No problem, thanks for your help with everything else though!

**Comment by epicfaace** [comment]
Hey @mxmlnkn -- I ended up modifying indexed_gzip to work with file objects without a fileno (https://github.com/pauldmccarthy/indexed_gzip/pull/55), so I'd like to add some PRs to ratarmount so that it doesn't crash with file objects that don't have a fileno. This wouldn't be a breaking change because it would just add support for file objects without a fileno for gzipped archive files. Let me know if you're OK with that!

**Comment by mxmlnkn** [comment]
As long as does not break any other behavior of ratarmount, I'm totally fine with that. I'll also try to review the MRs as fast as possible, but as you might have noticed, I currently can't find the time to add the support myself.

**Comment by epicfaace** [comment]
@mxmlnkn hey, whenever you get time to review -- I believe I've sent you the right PRs to enable file object support and get the Python API to work without needing fuse. Each PR builds on the other, in order, so when you do review, it would be best to review in this order:
- https://github.com/mxmlnkn/ratarmount/pull/59
- https://github.com/mxmlnkn/ratarmount/pull/62
- https://github.com/mxmlnkn/ratarmount/pull/63

**Comment by danieldjewell** [comment]
I would +1 this -- it does seem to be working with your original suggestions @mxmlnkn ... the part that is still a little funky is actually reading data: ```.read()``` requires the size and offset.

It would be great to have the ability to:
* Return all data for a given path (like zipfile does) ...
* Return an io stream/file-like object that would allow for compatibility with other libraries that are expecting to read from a file pointer/file-like object. (Not sure of the best way to implement this -- maybe loading the requested file into an ```io.BytesIO``` object?)

**Comment by mxmlnkn** [comment]
Guten Abend Daniel :),

As four your suggestions for the Python-API, could you take a look at the develop branch and see if the [`MountSource` interface](https://github.com/mxmlnkn/ratarmount/blob/develop/ratarmount.py#L390-L431) looks good to you? 

I began with that refactor while working on #64 and #65, that's why it isn't mentioned in this issue yet. These three issues are correlated to some extent.

**Comment by danieldjewell** [comment]
> Guten Abend Daniel :),
:grin:
 
> As four your suggestions for the Python-API, could you take a look at the develop branch and see if the [`MountSource` interface](https://github.com/mxmlnkn/ratarmount/blob/develop/ratarmount.py#L390-L431) looks good to you?

Yes, I think that's definitely more in the direction, although it would be excellent to have the option of using the SQLite indexing to speed access. 
 
> That interface adds an `open` function and a `listDir` function. I think these two are sufficient. The `read` function still exists and still requires the offset because it might be faster to read without having to return a temporary file object for internal use with ratarmount.

Would a temporary file object be necessary? I'm assuming (after a quick read of the code) that all of the caching/info required to make FUSE work could be utilized to make `read()` only require a path? 

My specific goal/use case is to be able to store data in a single file compressed (indexed tar) and then be able to read that data using the normal file-like/fp interfaces.

So instead of:
```python
file = open('somefile.json', 'r')
data = json.load(file)
``` 

Something like:

```python
ratar = ratarmount.open('archive.tar.zst', 'r')
#taking advantage of the indexing features!
data = json.load(ratar.file('/somefile.json'))
#and/or
data = json.loads(ratar.read('/somefile.json'))
```

Even Python's `ZipFile` library doesn't really have the ability to wrap a decompressed file in a seekable file-like container... (You have to manually do a ```bio = io.BytesIO(zipfile.read('filename'))``` which I suppose is OK too.) The only reason a wrapped file-like object would be nice is because some libraries are a little picky about reading -- many (like `pandas`) are pretty flexible: all/nearly all of the `pd.read_*` methods allow for a file name, a file buffer/fp, or a string as the input. But not all libraries are that flexible... some absolutely will not read a bytes object. (But then again, wrapping it by hand isn't *that* big of a deal.)

 

> I began with that refactor while working on #64 and #65, that's why it isn't mentioned in this issue yet. These three issues are correlated to some extent.

Kein problem! I just wanted to mention it since it seemed like a good time. 

(And as to why I don't just use a ZipFile .... Yes, it's widely supported but the compression is just so 1990s... and (unless I've missed out) you don't get the ability to store *NIX-y file data like you can with TAR...)

**Comment by JuniorJPDJ** [comment]
> Even Python's ZipFile library doesn't really have the ability to wrap a decompressed file in a seekable file-like container...

This is not a true. ZipFile allows seeking for a long time now.
Mine PR makes seeking on uncompressed file even faster.
python/cpython/pull/26227

> you don't get the ability to store *NIX-y file data like you can with TAR...)

You have this ability in ZIP, I'm not sure if py's ZipFile supports that extension tho.

**Comment by danieldjewell** [comment]
> > Even Python's ZipFile library doesn't really have the ability to wrap a decompressed file in a seekable file-like container...
>
> This is not a true. ZipFile allows seeking for a long time now.
> Mine PR makes seeking on uncompressed file even faster.
> [python/cpython/pull/26227](https://github.com/python/cpython/pull/26227)

But there's no way to have ZipFile (or ExtZipFile) return a file handle *for a file within the ZIP* is there? (I'm not talking about the file handle holding the ZIP Archive)

As in something like:
```python
zf = zipfile.open('ABC.zip','r')
#access a file handle *inside* the ZIP
file = zf.open('data.csv', 'r')
#file is now a normal file like object with read() seek() tell() etc
```

I don't think the above is possible?
The closest would be (would it not?):
```python
zf = zipfile.open('ABC.zip','r')
file = io.BytesIO(zf.read('data.csv'))
```

> > you don't get the ability to store *NIX-y file data like you can with TAR...)
> 
> You have this ability in ZIP, I'm not sure if py's ZipFile supports that extension tho.

Oh, hmmm. News to me. I usually just unzip things when on Linux/Unix :grin: good to know, thanks!

**Comment by JuniorJPDJ** [comment]
> But there's no way to have ZipFile (or ExtZipFile) return a file handle for a file within the ZIP is there? (I'm not talking about the file handle holding the ZIP Archive)

Yes. There is ;)

It works exactly how you mentioned with `file`.

**Comment by JuniorJPDJ** [comment]
There are even some FUSE filesystem using this.
eg. https://github.com/openscopeproject/ZipROFS

**Comment by danieldjewell** [comment]
> > But there's no way to have ZipFile (or ExtZipFile) return a file handle for a file within the ZIP is there? (I'm not talking about the file handle holding the ZIP Archive)
> 
> Yes. There is ;)
> 
> It works exactly how you mentioned with `file`.

Sure enough... there it is! https://github.com/python/cpython/blob/f7635f0e542c916bfd62542e7b60ee23ff681d0d/Lib/zipfile.py#L1466

Cool! Thanks for pointing this out. 

It's actually kinda cool how it works ... returns a `zipfile.ZipExtFile` which is a subclass of `io.BufferedIOBase`... I wonder how much of the entire zip file gets read/decompressed when reading a single file...

**Comment by mxmlnkn** [comment]
> > As four your suggestions for the Python-API, could you take a look at the develop branch and see if the [`MountSource` interface](https://github.com/mxmlnkn/ratarmount/blob/develop/ratarmount.py#L390-L431) looks good to you?
> 
> Yes, I think that's definitely more in the direction, although it would be excellent to have the option of using the SQLite indexing to speed access.

That should already be the case. `MountSource` is only an interface. `SQLiteIndexTar` derives from it and, as the name contains, it tries to load an index file to speed up access. There also is a generic [`ratarmount.openMountSource`](https://github.com/mxmlnkn/ratarmount/blob/develop/ratarmount.py#L2348) function which automatically chooses `SQLiteIndexTar` or a suitable class of the other `MountSource`-derived classes for any given file path. Maybe I should rename `openMountSource` into `open`...

> > That interface adds an `open` function and a `listDir` function. I think these two are sufficient. The `read` function still exists and still requires the offset because it might be faster to read without having to return a temporary file object for internal use with ratarmount.
> 
> Would a temporary file object be necessary? I'm assuming (after a quick read of the code) that all of the caching/info required to make FUSE work could be utilized to make `read()` only require a path?

Generally, no. When using the `SQLiteIndexedTar` backend no temporary file is created anywhere. What you get from the `open` call is a very lightweight Python file object, which has a pointer to the underlying TAR file as well as the offset and size for that file inside the TAR and a separate offset inside the file itself. Nothing is explicitly extracted to `/tmp` or to memory. (The yet unreleased rar backend based on the `rarfile` module seems to extract files into `/tmp` when they are opened throught eh Python-API)

> My specific goal/use case is to be able to store data in a single file compressed (indexed tar) and then be able to read that data using the normal file-like/fp interfaces.

You don't need a tar for that. Ratarmount is also able to open simple `.bz2`, `.gz`, `.xz`, and `.zst` files with the seeking support. However, the integration with `openMountSource` might be a bit weird because even though it is a simple compressed file, it will return an SQLiteIndexedTar file instead of a file object directly. However, returning a file object in some cases and a MountSource in other cases might also be confusing behavior :/. It's hard to decide. So, given your example, you would call it in the develop branch with:

```python
ratar = ratarmount.openMountSource('archive.tar.zst', 'r')
data = json.load(ratar.open(ratar.getFileInfo('/somefile.json')))
#and/or
df = pd.read_csv(ratar.read(ratar.getFileInfo('/somefile.csv')))
```

The two-stepped call with `getFileInfo` and then `MountSource.open` is necessary because a simple file name is not unique as there might be multiple files with the same name/path in the archive. And I didn't want to add the `fileVersion` parameter to all other API functions, so you first have to create a kind of unique identifier with the `getFileInfo` method.

**Comment by JuniorJPDJ** [comment]
> I wonder how much of the entire zip file gets read/decompressed when reading a single file...

ZIP files have each file inside compressed separatly.
`zipfile` decompresses only one file when accessing one file and decompresses everything (of in-zip file) before the place you seek to.
Example:
you have zip with tree like:
|
|- file0 (100MB)
|- file1 (2MB)
|- file2 (1MB)
when you do `f = zip.open('file1')` and then `f.seek(300)` it decompresses all 300 bytes so it can seek, but doesn't touch `file0` at all

**Comment by GodOfSmallThings** [comment]
I wanted to access to the files in tar, so I did like this and the file was successfully loaded.
`
ratar = ratarmount.openMountSource("foo.tar")
txt = ratar.open(ratar.getFileInfo('texts/bar.txt")) 
`

However, I need to access to files in many tars randomly, and want to use like this:
'
ratar = ratarmount.openMountSource("folder")
txt1 = ratar.open(ratar.getFileInfo("foo1.tar/texts/bar1.txt"))
txt2 = ratar.open(ratar.getFileInfo("foo2.tar/texts/bar2.txt"))
txt3 = ratar.open(ratar.getFileInfo("foo3.tar/texts/bar3.txt"))
'
Of course this isn't working. To explain more specifically, I need to make a dictionary and access to files fast with its keys.
Is there a way to access files in multiple tars like that?

**Comment by mxmlnkn** [comment]
@GodOfSmallThings Thank so much for testing this bleeding edge develop branch! That gives me some things to improve on.

There are multiple issues:

 - `getFileInfo` expects the path to begin with a leading `/` in some cases. I'll try to take care to make it also work in all cases without the leading slash.
 - `openMountSource` simply opens the folder. If you also want to mount the archives inside the folder, then you need to specify the `recursive` option just like with the CLI. I won't change that behavior. However, `ratarmount.openMountSource("folder/", recursive = True)` will not work even though I think it should work. I'll fix this soon! For now there is a workaround you can use instead, see below.

In the end, this should work with the version you already have installed:

```python3
import ratarmount
ratar = ratarmount.AutoMountLayer(ratarmount.openMountSource("folder"), recursive = True)
print(ratar.listDir("/"))
print(ratar.listDir("/foo1.tar"))
txt1 = ratar.open(ratar.getFileInfo("/foo1.tar/texts/bar1.txt"))
```

Personally, I found it confusing that `AutoMountLayer` was not named `AutoMountSource`. I might change that.

I hope this answers your question.


Notes for myself / issues I found while testing this:

 - The error message when using `ratar.open(string)` instead of `ratar.open(fileInfo)` is not helpful at all. Make it more helpful or make it simply work as an alternative interface even if no file version can be specified through this.
 - `AutoMountLayer.open` fails with an assertion that the size of a stenciledFile must be larger than zero when opening a symbolic link. I'm not sure whether to automatically resolve links or whether to just improve the error message. Also, I fear that the assert will also be triggered for regular files which are simply empty.

**Comment by GodOfSmallThings** [comment]
@mxmlnkn 

Thank you very much! I tested it and it works just fine.

It became bit slower if I access file like `txt1 = ratar.open(ratar.getFileInfo("/foo1.tar/texts/bar1.txt"))`, but it's way better than opening openMountSource every time I access files in different tar files. 

In my case, processing time through tar in folder for each file was 3x~8x slower than directly accessing to the file from this below code. (0.00015 -> 0.00035~0.0011 sec)
```
ratar = ratarmount.openMountSource("foo.tar")
txt = ratar.open(ratar.getFileInfo('texts/bar.txt")) 
```

**Comment by mxmlnkn** [comment]
@GodOfSmallThings Interesting to know that it became slower. Probably because it has to look up to which archive the path belongs. But this should only affect the `open` call. After you have the file object, the calls to `read` and `seek` should be equally fast. `open` calls are known to be slow in many programming languages and should be used with caution anyway. Therefore, improving speed does not have much of a priority for me but I might take a quick lock.

**Comment by GodOfSmallThings** [comment]
@mxmlnkn 

This much of slowdown is also no problem for me, I was just reporting phenomenon I saw.
Thank you very much for help.

**Comment by mxmlnkn** [comment]
This issue got quite long. I think, I fixed most of the issues mentioned here including the original issue of better file-object support when given to SQLiteIndexedTar. I did take a double look at the changes already mentioned above and pushed them to master and will release v0.9.0 today.

I even added support for file-like Python object without a valid fileno to `indexed_bzip2` using the Python C API just like @epicfaace did for `indexed_gzip`. This enables recursive mounting over different archive boundaries whereas it only worked for uncompressed TAR files inside other TAR files. The only compression backend missing that support would be the `indexed_zstd` module. The FuseMount class is now only a thin wrapper around fuse.py and the only step left for real "use as a library" support especially without FUSE would be to implement #65.

Small overview over the current architecture:

![MountSource](https://user-images.githubusercontent.com/6842824/133676334-1b5ea3b1-93f7-4369-a009-8f76e0459f7c.png)
